### ISSUE TITLE
Hide plugins (in structure mode) - feature [re]quest(ion)

### ISSUE THREAD

**Issue by creimers** [django-cms/django-cms#4512]
Other CMS provide the user with the possibility to 'hide' plugins, i.e. to not show them to the ordinary site visitor without having to delete them.

Is a similiar functionality in the making for djangoCMS by chance?

This might have been asked here before, in which case I apologize for my insufficient research.

**Comment by yakky** [comment]
What's the use case for this feature? (BTW you can easily create a ``HiddenPluginsPlugin`` which hides the contained plugins according to some specific rules -like: hide plugins for non staff users)

**Comment by creimers** [comment]
Use case: Some banner or image on a page that you periodically want to manually hide and show. Or some content that you need later but now happen to have the time to insert it in the CMS. You could then leave it hidden for the time being.

Can't think of anything more illustrating right now, but I have been asked this by multiple clients who were switching *to* djangoCMS and appreciated such functionality in their old CMS.

Of course I could create a ``HiddenPluginsPlugin``, but a 'hide/show' option along with 'copy', 'edit', 'delete', ... in the structure view would appear much more convenient to me.

**Comment by yakky** [comment]
Looks legit. Thanks for the suggestion @creimers

**Comment by creimers** [comment]
@yakky Yes I'll be able to contribute this with some help. Is there documentation on how the structure mode works/is built?

**Comment by jrief** [comment]
You may have a look at the segmentation plugin.
https://github.com/aldryn/aldryn-segmentation

djangocms-casade also offers segmentation.

**Comment by yakky** [comment]
@creimers no need for structure mode: https://github.com/divio/django-cms/blob/develop/cms/plugin_base.py#L333-L340 use these to add items to the placeholder / plugin menu, a view in ``PlaceholderAdminMixin``, then I guess we could handle this with a field that saves whether to hide the plugin or not.
In edit mode the plugin must be rendered (eventually adding a specific class to make visible they are "hidden"), so you'll need to adjust the ``get_plugins`` function https://github.com/divio/django-cms/blob/develop/cms/utils/plugins.py#L26

**Comment by evildmp** [comment]
This has been discussed before, in fact I discussed it recently with someone but I can't remember where or with whom.

In short, it's a feature I've wished for since 2.3, and it would have enormous practical and time-saving use.

Each plugin that is ``publish``-enabled (this might not be appropriate for every kind of plugin) should have a checkbox, ``Publish``, that is on by default.

Whether this control should be visible in content as well as structure mode is not clear; why not both, if it's easy enough?

Actual use cases:

* there was an emergency situation, and the home page of the site needed a prominent message, but we had to wait for the go-ahead before publishing it. Over the next few days we needed sometimes no message, and sometimes a modified version of it.

* every so often, we want not to display a certain section of a page temporarily

* on one page, we regularly published the same announcement at different times of the year, for a week or so; the rest of the time it wasn't needed

 ``publish`` control over plugins would make managing this much easier, and without repeating our efforts

So, very strongly in favour.

**Comment by nmashton** [comment]
My client has also requested this feature for the purpose of showing and hiding seasonal content, so I'm strongly in favor as well.

How should this feature interact with the new Alias plugin? Should an Alias that points to a hidden plugin always be hidden, or should its visibility always be independent of the source plugin? (I assume that if the source plugin is visible, you should be able to hide the Alias if you want to.)

**Comment by benzkji** [comment]
implementing this on your own is not that hard, and indeed very usefull...I'm using this since quite some time. using arbitrary 3rd party plugins is more difficult, though.

Proof of concept: 
https://github.com/bnzk/djangocms-baseplugins/blob/develop/djangocms_baseplugins/baseplugin/models.py#L80  
https://github.com/bnzk/djangocms-baseplugins/blob/develop/djangocms_baseplugins/baseplugin/cms_plugins.py#L20

Making it opaque also works: https://github.com/bnzk/djangocms-baseplugins/blob/develop/djangocms_baseplugins/baseplugin/templates/djangocms_baseplugins/baseplugin_styles.html

Adding context menu items (show/hide) is a thought...but not a priority.

**Comment by jrief** [comment]
djangocms-cascade has such a feature implemented. There you can hide a plugin and all of its children.
